File: java/src/main/java/com/b4n1/web/AgentBrowser.java
package com.b4n1.web;

import java.io.BufferedReader;
import java.io.File;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

/**
 * B4n1Web Agent Browser.
 *
 * A browser instance optimized for AI agent workflows.
 * Requires B4n1Web binary to be installed.
 *
 * <pre>{@code
 * var browser = new AgentBrowser(BrowserMode.LIGHT);
 * var page = browser.goto("https://example.com");
 * System.out.println(page.getMarkdown());
 * browser.close();
 * }</pre>
 */
public class AgentBrowser implements AutoCloseable {
    private static final String SDK_VERSION = "0.7.0";
    private final BrowserOptions options;
    private final String binaryPath;

    public AgentBrowser() {
        this(new BrowserOptions());
    }

    public AgentBrowser(BrowserOptions options) {
        this.options = Objects.requireNonNull(options, "options cannot be null");

        String path = findBinary();
        if (path == null || path.isEmpty()) {
            throw new BinaryNotFoundException();
        }
        this.binaryPath = path;

        checkVersionCompatibility();
    }

    /**
     * Check if binary version matches SDK version.
     * Prints warning to stderr if mismatch detected.
     */
    private static void checkVersionCompatibility() {
        String binaryVersion = getVersion();
        if (binaryVersion.equals("unknown")) {
            return;
        }

        if (!binaryVersion.equals(SDK_VERSION)) {
            System.err.println("Version mismatch: SDK v" + SDK_VERSION + " requires binary v" + SDK_VERSION +
                ", but found v" + binaryVersion + ". Some features may not work correctly.");
        }
    }

    /**
     * Navigate to a URL and extract structured content.
     */
    public Page goto_(String url) throws NavigationException {
        return goto_(url, null);
    }

    /**
     * Navigate to a URL with an optional CSS selector to wait for.
     */
    public Page goto_(String url, String waitFor) throws NavigationException {
        try {
            List<String> cmd = new ArrayList<>();
            cmd.add(binaryPath);
            cmd.add("goto");
            cmd.add(url);
            cmd.add("--mode");
            cmd.add(options.getMode().getValue());
            if (waitFor != null && !waitFor.isEmpty()) {
                cmd.add("--wait-for");
                cmd.add(waitFor);
            }

            ProcessBuilder pb = new ProcessBuilder(cmd);
            pb.redirectErrorStream(true);
            Process process = pb.start();

            String output = new BufferedReader(new InputStreamReader(process.getInputStream()))
                    .lines()
                    .collect(Collectors.joining("\n"));

            boolean finished = process.waitFor(options.getTimeout(), java.util.concurrent.TimeUnit.SECONDS);
            if (!finished) {
                process.destroyForcibly();
                throw new NavigationException(url, "Binary timed out after " + options.getTimeout() + "s");
            }

            if (process.exitValue() != 0) {
                throw new NavigationException(url, "Binary error: exit code " + process.exitValue());
            }

            return parseOutput(url, output);

        } catch (NavigationException e) {
            throw e;
        } catch (Exception e) {
            throw new NavigationException(url, e);
        }
    }

    /**
     * Navigate to URL (alias for goto to avoid reserved keyword).
     */
    public Page navigate(String url) {
        return goto_(url);
    }

    /**
     * Take a screenshot of the current page with specified dimensions.
     */
    public String screenshot(int width, int height) {
        try {
            ProcessBuilder pb = new ProcessBuilder(
                binaryPath, "screenshot", "--width", String.valueOf(width), "--height", String.valueOf(height),
                "--mode", options.getMode().getValue()
            );
            pb.redirectErrorStream(true);
            Process process = pb.start();

            String output = new BufferedReader(new InputStreamReader(process.getInputStream()))
                    .lines()
                    .collect(Collectors.joining("\n"));

            process.waitFor(options.getTimeout(), java.util.concurrent.TimeUnit.SECONDS);
            return output.trim();
        } catch (Exception e) {
            throw new RuntimeException("Screenshot failed", e);
        }
    }

    /**
     * Wait for a CSS selector to appear within the given timeout.
     */
    public boolean waitForSelector(String selector, int timeoutMs) {
        try {
            ProcessBuilder pb = new ProcessBuilder(
                binaryPath, "wait-selector", selector, "--timeout", String.valueOf(timeoutMs),
                "--mode", options.getMode().getValue()
            );
            pb.redirectErrorStream(true);
            Process process = pb.start();

            String output = new BufferedReader(new InputStreamReader(process.getInputStream()))
                    .lines()
                    .collect(Collectors.joining("\n"));

            process.waitFor(timeoutMs / 1000 + 1, java.util.concurrent.TimeUnit.SECONDS);
            return process.exitValue() == 0;
        } catch (Exception e) {
            return false;
        }
    }

    /**
     * Click an element matching the given CSS selector.
     */
    public void click(String selector) {
        try {
            ProcessBuilder pb = new ProcessBuilder(
                binaryPath, "click", selector, "--mode", options.getMode().getValue()
            );
            pb.redirectErrorStream(true);
            Process process = pb.start();
            process.waitFor(options.getTimeout(), java.util.concurrent.TimeUnit.SECONDS);
        } catch (Exception e) {
            throw new RuntimeException("Click failed for selector: " + selector, e);
        }
    }

    /**
     * Type text into an element matching the given CSS selector.
     */
    public void typeText(String selector, String text, boolean clearFirst) {
        try {
            List<String> cmd = new ArrayList<>();
            cmd.add(binaryPath);
            cmd.add("type");
            cmd.add(selector);
            cmd.add(text);
            cmd.add("--mode");
            cmd.add(options.getMode().getValue());
            if (clearFirst) {
                cmd.add("--clear-first");
            }

            ProcessBuilder pb = new ProcessBuilder(cmd);
            pb.redirectErrorStream(true);
            Process process = pb.start();
            process.waitFor(options.getTimeout(), java.util.concurrent.TimeUnit.SECONDS);
        } catch (Exception e) {
            throw new RuntimeException("TypeText failed for selector: " + selector, e);
        }
    }

    /**
     * Convenience: fetch links directly from a URL.
     */
    public static String[] getLinksFromPage(String url) {
        try (AgentBrowser browser = new AgentBrowser()) {
            Page page = browser.goto_(url);
            return page.getLinksArray();
        }
    }

    private Page parseOutput(String url, String output) {
        StringBuilder markdown = new StringBuilder();
        List<String> links = new ArrayList<>();
        String jsOutput = null;

        for (String line : output.split("\n")) {
            if (line.startsWith("URL:")) {
                continue;
            } else if (line.startsWith("Markdown:")) {
                continue;
            } else if (line.startsWith("Links:")) {
                try {
                    String linksJson = line.substring(6).trim();
                    links = parseLinksJson(linksJson);
                } catch (Exception e) {
                    links = new ArrayList<>();
                }
            } else if (line.startsWith("JSOutput:")) {
                jsOutput = line.substring(9).trim();
            } else {
                if (markdown.length() > 0) {
                    markdown.append("\n");
                }
                markdown.append(line);
            }
        }

        Page page = new Page();
        page.setUrl(url);
        page.setMarkdown(markdown.toString().trim());
        page.setLinks(links);
        page.setJsOutput(jsOutput);
        return page;
    }

    private List<String> parseLinksJson(String json) {
        if (json == null || json.isEmpty() || json.equals("[]")) {
            return new ArrayList<>();
        }
        json = json.trim();
        if (!json.startsWith("[")) {
            return new ArrayList<>();
        }
        List<String> result = new ArrayList<>();
        json = json.substring(1, json.length() - 1);
        for (String item : json.split(",")) {
            item = item.trim();
            if (item.startsWith("'") && item.endsWith("'")) {
                result.add(item.substring(1, item.length() - 1));
            } else if (item.startsWith("\"") && item.endsWith("\"")) {
                result.add(item.substring(1, item.length() - 1));
            }
        }
        return result;
    }

    private String findBinary() {
        try {
            String resourcePath = "/native/linux-x86_64/b4n1web-linux";
            java.net.URL url = getClass().getResource(resourcePath);
            if (url != null) {
                File bundledBinary = extractBundledBinary(resourcePath);
                if (bundledBinary != null && bundledBinary.canExecute()) {
                    return bundledBinary.getAbsolutePath();
                }
            }
        } catch (Exception e) {
        }

        String[] possiblePaths = {
            "/usr/local/bin/b4n1web",
            "/usr/bin/b4n1web",
            System.getProperty("user.home") + "/.local/bin/b4n1web",
            System.getProperty("user.home") + "/.b4n1web/bin/b4n1web"
        };

        for (String path : possiblePaths) {
            File f = new File(path);
            if (f.exists() && f.canExecute()) {
                return path;
            }
        }
        return null;
    }

    private File extractBundledBinary(String resourcePath) {
        try {
            File tempDir = new File(System.getProperty("java.io.tmpdir"), "b4n1web");
            tempDir.mkdirs();
            File tempBinary = new File(tempDir, "b4n1web");

            try (java.io.InputStream in = getClass().getResourceAsStream(resourcePath);
                 java.io.FileOutputStream out = new java.io.FileOutputStream(tempBinary)) {
                if (in == null) return null;
                byte[] buffer = new byte[8192];
                int bytesRead;
                while ((bytesRead = in.read(buffer)) != -1) {
                    out.write(buffer, 0, bytesRead);
                }
            }
            tempBinary.setExecutable(true);
            return tempBinary;
        } catch (Exception e) {
            return null;
        }
    }

    /**
     * Get B4n1Web binary version.
     */
    public static String getVersion() {
        String[] paths = {
            "/usr/local/bin/b4n1web",
            "/usr/bin/b4n1web",
            System.getProperty("user.home") + "/.local/bin/b4n1web",
            System.getProperty("user.home") + "/.b4n1web/bin/b4n1web"
        };

        for (String path : paths) {
            File f = new File(path);
            if (f.exists() && f.canExecute()) {
                try {
                    Process pb = new ProcessBuilder(path, "--version").start();
                    String version = new BufferedReader(new InputStreamReader(pb.getInputStream()))
                            .lines()
                            .collect(Collectors.joining("\n"));
                    pb.waitFor(5, java.util.concurrent.TimeUnit.SECONDS);
                    return version.trim();
                } catch (Exception e) {
                }
            }
        }
        return "unknown";
    }

    @Override
    public void close() {
    }
}

File: csharp/src/AgentBrowser.cs
using System.Diagnostics;

namespace B4N1Web;

/// <summary>
/// B4n1Web Agent Browser
/// 
/// A browser instance optimized for AI agent workflows.
/// Requires B4n1Web binary to be installed.
/// </summary>
/// <example>
/// <code>
/// using B4N1Web;
/// 
/// var browser = new AgentBrowser(new BrowserOptions { Mode = BrowserMode.Light });
/// var page = await browser.GotoAsync("https://example.com");
/// Console.WriteLine(page.Markdown);
/// browser.Close();
/// </code>
/// </example>
public class AgentBrowser : IDisposable
{
    private const string SdkVersion = "0.5.0";
    private readonly BrowserOptions _options;
    private readonly string _binaryPath;
    private Page? _lastPage;
    private bool _disposed;

    public AgentBrowser(BrowserOptions? options = null)
    {
        _options = options ?? new BrowserOptions();

        _binaryPath = FindBinary();
        if (string.IsNullOrEmpty(_binaryPath))
        {
            throw new BinaryNotFoundException();
        }

        // Check version compatibility (non-fatal warning)
        CheckVersionCompatibility();
    }

    /// <summary>
    /// Check if binary version matches SDK version.
    /// Prints warning to stderr if mismatch detected.
    /// </summary>
    private static void CheckVersionCompatibility()
    {
        var binaryVersion = GetVersion();
        if (binaryVersion == "unknown")
        {
            return;
        }

        if (binaryVersion != SdkVersion)
        {
            Console.Error.WriteLine(
                $"⚠️  Version mismatch: SDK v{SdkVersion} requires binary v{SdkVersion}, " +
                $"but found v{binaryVersion}. Some features may not work correctly.");
        }
    }

    /// <summary>
    /// Navigate to a URL and extract structured content
    /// </summary>
    public async Task<Page> GotoAsync(string url, string? waitFor = null)
    {
        var args = $"goto {url} --mode {_options.Mode.ToString().ToLower()}";
        if (waitFor != null)
        {
            args += $" --wait-for \"{waitFor}\"";
        }

        var startInfo = new ProcessStartInfo
        {
            FileName = _binaryPath,
            Arguments = args,
            UseShellExecute = false,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            CreateNoWindow = true
        };

        using var process = new Process { StartInfo = startInfo };
        process.Start();

        var output = await process.StandardOutput.ReadToEndAsync();
        var error = await process.StandardError.ReadToEndAsync();
        
        var timedOut = await Task.Run(() => process.WaitForExit(_options.Timeout * 1000));
        
        if (!timedOut)
        {
            process.Kill();
            throw new Exception($"Binary timed out after {_options.Timeout}s");
        }

        if (process.ExitCode != 0)
        {
            throw new Exception($"Binary error: {error}");
        }

        var page = ParseOutput(url, output);
        _lastPage = page;
        return page;
    }

    /// <summary>
    /// Navigate to URL (synchronous version)
    /// </summary>
    public Page Goto(string url, string? waitFor = null)
    {
        return GotoAsync(url, waitFor).GetAwaiter().GetResult();
    }

    /// <summary>
    /// Take a screenshot of the current page
    /// </summary>
    public string Screenshot(int width, int height)
    {
        if (_lastPage == null)
            throw new InvalidOperationException("No page loaded. Call Goto or GotoAsync first.");

        var startInfo = new ProcessStartInfo
        {
            FileName = _binaryPath,
            Arguments = $"screenshot --url {_lastPage.Url} --width {width} --height {height}",
            UseShellExecute = false,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            CreateNoWindow = true
        };

        using var process = new Process { StartInfo = startInfo };
        process.Start();

        var output = process.StandardOutput.ReadToEnd();
        var error = process.StandardError.ReadToEnd();

        var timedOut = process.WaitForExit(_options.Timeout * 1000);

        if (!timedOut)
        {
            process.Kill();
            throw new Exception($"Binary timed out after {_options.Timeout}s");
        }

        if (process.ExitCode != 0)
        {
            throw new Exception($"Binary error: {error}");
        }

        foreach (var line in output.Split('\n'))
        {
            if (line.StartsWith("Screenshot:"))
            {
                var b64 = line[11..].Trim();
                if (!string.IsNullOrEmpty(b64))
                    return b64;
            }
        }

        throw new Exception("No screenshot data returned from binary");
    }

    /// <summary>
    /// Wait for a CSS selector to appear on the page
    /// </summary>
    public bool WaitForSelector(string selector, int timeoutMs)
    {
        if (_lastPage == null)
            throw new InvalidOperationException("No page loaded. Call Goto or GotoAsync first.");

        var startInfo = new ProcessStartInfo
        {
            FileName = _binaryPath,
            Arguments = $"wait-for-selector \"{selector}\" --url {_lastPage.Url} --timeout {timeoutMs}",
            UseShellExecute = false,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            CreateNoWindow = true
        };

        using var process = new Process { StartInfo = startInfo };
        process.Start();

        var output = process.StandardOutput.ReadToEnd();
        var error = process.StandardError.ReadToEnd();

        var timedOut = process.WaitForExit(_options.Timeout * 1000);

        if (!timedOut)
        {
            process.Kill();
            throw new Exception($"Binary timed out after {_options.Timeout}s");
        }

        if (process.ExitCode != 0)
        {
            throw new Exception($"Binary error: {error}");
        }

        foreach (var line in output.Split('\n'))
        {
            if (line.StartsWith("Found:"))
            {
                return line[6..].Trim() == "true";
            }
        }

        return false;
    }

    /// <summary>
    /// Click on an element by CSS selector
    /// </summary>
    public void Click(string selector)
    {
        if (_lastPage == null)
            throw new InvalidOperationException("No page loaded. Call Goto or GotoAsync first.");

        var startInfo = new ProcessStartInfo
        {
            FileName = _binaryPath,
            Arguments = $"click \"{selector}\" --url {_lastPage.Url}",
            UseShellExecute = false,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            CreateNoWindow = true
        };

        using var process = new Process { StartInfo = startInfo };
        process.Start();

        var error = process.StandardError.ReadToEnd();
        var timedOut = process.WaitForExit(_options.Timeout * 1000);

        if (!timedOut)
        {
            process.Kill();
            throw new Exception($"Binary timed out after {_options.Timeout}s");
        }

        if (process.ExitCode != 0)
        {
            throw new Exception($"Binary error: {error}");
        }
    }

    /// <summary>
    /// Type text into an element by CSS selector
    /// </summary>
    public void TypeText(string selector, string text, bool clearFirst)
    {
        if (_lastPage == null)
            throw new InvalidOperationException("No page loaded. Call Goto or GotoAsync first.");

        var clearArg = clearFirst ? " --clear-first" : "";
        var startInfo = new ProcessStartInfo
        {
            FileName = _binaryPath,
            Arguments = $"type-text \"{selector}\" \"{text}\" --url {_lastPage.Url}{clearArg}",
            UseShellExecute = false,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            CreateNoWindow = true
        };

        using var process = new Process { StartInfo = startInfo };
        process.Start();

        var error = process.StandardError.ReadToEnd();
        var timedOut = process.WaitForExit(_options.Timeout * 1000);

        if (!timedOut)
        {
            process.Kill();
            throw new Exception($"Binary timed out after {_options.Timeout}s");
        }

        if (process.ExitCode != 0)
        {
            throw new Exception($"Binary error: {error}");
        }
    }

    /// <summary>
    /// Get links from the last visited page
    /// </summary>
    public string[] GetLinks()
    {
        return _lastPage?.Links.ToArray() ?? Array.Empty<string>();
    }

    /// <summary>
    /// Fetch links from a URL without creating a browser instance (static)
    /// </summary>
    public static string[] GetLinksFromPage(string url)
    {
        var binaryPath = FindBinary();
        if (string.IsNullOrEmpty(binaryPath))
            throw new BinaryNotFoundException();

        var startInfo = new ProcessStartInfo
        {
            FileName = binaryPath,
            Arguments = $"goto {url} --mode light",
            UseShellExecute = false,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            CreateNoWindow = true
        };

        using var process = new Process { StartInfo = startInfo };
        process.Start();

        var output = process.StandardOutput.ReadToEnd();
        process.WaitForExit(30000);

        if (process.ExitCode != 0)
            return Array.Empty<string>();

        foreach (var line in output.Split('\n'))
        {
            if (line.StartsWith("Links:"))
            {
                try
                {
                    return System.Text.Json.JsonSerializer.Deserialize<List<string>>(line[6..])
                        ?.ToArray() ?? Array.Empty<string>();
                }
                catch
                {
                    return Array.Empty<string>();
                }
            }
        }

        return Array.Empty<string>();
    }

    /// <summary>
    /// Parse text output from the binary
    /// </summary>
    private Page ParseOutput(string url, string output)
    {
        var markdown = new List<string>();
        var links = new List<string>();
        string? screenshot = null;
        string? jsOutput = null;

        foreach (var line in output.Split('\n'))
        {
            if (line.StartsWith("URL:"))
            {
                continue;
            }
            else if (line.StartsWith("Markdown:"))
            {
                continue;
            }
            else if (line.StartsWith("Links:"))
            {
                try
                {
                    links = System.Text.Json.JsonSerializer.Deserialize<List<string>>(line[6..]) 
                        ?? new List<string>();
                }
                catch
                {
                    links = new List<string>();
                }
            }
            else if (line.StartsWith("Screenshot:"))
            {
                screenshot = line[11..].Trim();
                if (string.IsNullOrEmpty(screenshot)) screenshot = null;
            }
            else if (line.StartsWith("JsOutput:"))
            {
                jsOutput = line[9..].Trim();
                if (string.IsNullOrEmpty(jsOutput)) jsOutput = null;
            }
            else
            {
                markdown.Add(line);
            }
        }

        return new Page
        {
            Url = url,
            Markdown = string.Join("\n", markdown).Trim(),
            Links = links,
            Screenshot = screenshot,
            JsOutput = jsOutput
        };
    }

    /// <summary>
    /// Find b4n1web binary in bundled location or system install
    /// </summary>
    private static string FindBinary()
    {
        // 1. Check bundled binary (bundled as embedded resource)
        var bundledPath = ExtractBundledBinary();
        if (!string.IsNullOrEmpty(bundledPath))
        {
            return bundledPath;
        }

        // 2. Check system install locations
        var possiblePaths = new[]
        {
            "/usr/local/bin/b4n1web",
            "/usr/bin/b4n1web",
            Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".local/bin/b4n1web"),
            Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".b4n1web/bin/b4n1web"),
        };

        foreach (var path in possiblePaths)
        {
            if (File.Exists(path))
            {
                return path;
            }
        }

        return string.Empty;
    }

    /// <summary>
    /// Extract bundled binary from embedded resources to temp directory
    /// </summary>
    private static string? ExtractBundledBinary()
    {
        try
        {
            var assembly = typeof(AgentBrowser).Assembly;
            var resourceName = "B4N1Web.native.linux-x64.b4n1web-linux";
            using var stream = assembly.GetManifestResourceStream(resourceName);
            if (stream == null) return null;

            var tempDir = Path.Combine(Path.GetTempPath(), "b4n1web");
            Directory.CreateDirectory(tempDir);
            var tempBinary = Path.Combine(tempDir, "b4n1web");

            using var fs = new FileStream(tempBinary, FileMode.Create, FileAccess.Write);
            stream.CopyTo(fs);

            // Make executable on Unix
            if (Environment.OSVersion.Platform == PlatformID.Unix || Environment.OSVersion.Platform == PlatformID.MacOSX)
            {
                var chmod = new Process { StartInfo = new ProcessStartInfo("chmod", $"755 {tempBinary}") { UseShellExecute = false } };
                chmod.Start();
                chmod.WaitForExit();
            }

            return tempBinary;
        }
        catch
        {
            return null;
        }
    }

    /// <summary>
    /// Get B4n1Web binary version
    /// </summary>
    public static string GetVersion()
    {
        var path = FindBinary();
        if (string.IsNullOrEmpty(path))
        {
            return "unknown";
        }

        try
        {
            var startInfo = new ProcessStartInfo
            {
                FileName = path,
                Arguments = "--version",
                UseShellExecute = false,
                RedirectStandardOutput = true,
                CreateNoWindow = true
            };

            using var process = new Process { StartInfo = startInfo };
            process.Start();
            var output = process.StandardOutput.ReadToEnd();
            process.WaitForExit(5000);
            
            return output.Trim();
        }
        catch
        {
            return "unknown";
        }
    }

    /// <summary>
    /// Close the browser session
    /// </summary>
    public void Close()
    {
        Dispose();
    }

    public void Dispose()
    {
        if (!_disposed)
        {
            _disposed = true;
        }
        GC.SuppressFinalize(this);
    }
}
